controller/kubernetes_endpoint_controller.go

```diff
@@ -0,0 +1,373 @@
+	// Retrieve a list of share manager pod by the Endpoint
+	shareManagerPods, err := c.ds.ListShareManagerPodsRO(endpoint.Name)
+	if err != nil {
+		return err
+	}
+
+	log := getLoggerForKubernetesEndpoint(c.logger, endpoint)
+
+	// A map of desired share manager IP and its pod.
+	desiredIPPod := make(map[string]*corev1.Pod)
+	for _, pod := range shareManagerPods {
+		storageIP := c.ds.GetStorageIPFromPod(pod)
+		if _, exist := desiredIPPod[storageIP]; exist {
+			log.Warnf("Found duplicated share manager pod storage IP %v", storageIP)
+			continue
+		}
+
+		desiredIPPod[storageIP] = pod
+	}
+
+	// A map of existing IP in the Endpoint and its endpoint address.
+	existingIPEndpointAddress := make(map[string]corev1.EndpointAddress)
+	for _, subset := range endpoint.Subsets {
+		for _, address := range subset.Addresses {
+			existingIPEndpointAddress[address.IP] = address
+		}
+	}
+
+	// Identify the address to delete, add and update.
+	ipSubsetToDelete := c.identifyIPSubsetToDelete(endpoint, desiredIPPod)
+	ipSubsetToAdd, addressToUpdate := c.identifyIPSubsetToAddOrUpdate(existingIPEndpointAddress, desiredIPPod, createDesiredSubsetForShareManager)
+
+	// If no change, log and return.
+	if len(ipSubsetToDelete) == 0 && len(ipSubsetToAdd) == 0 && len(addressToUpdate) == 0 {
+		log.Trace("No need to reconcile Kubernetes Endpoint because there is no change to the subset")
+		return nil
+	}
+
+	// Update the Endpoint subsets with the identified changes.
+	updateEndpointSubsets(endpoint, ipSubsetToDelete, ipSubsetToAdd, addressToUpdate, log)
+
+	return nil
+}
+
+func (c *KubernetesEndpointController) identifyIPSubsetToDelete(endpoint *corev1.Endpoints, desiredIPPod map[string]*corev1.Pod) map[string]bool {
+	addressToDelete := make(map[string]bool)
+	for _, subset := range endpoint.Subsets {
+		for _, address := range subset.Addresses {
+			// Check if the address is in the desired IP pod map.
+			if _, exists := desiredIPPod[address.IP]; !exists {
+				// If it does not exist, mark it for deletion.
+				addressToDelete[address.IP] = true
+			}
+		}
+	}
+	return addressToDelete
+}
+
+func (c *KubernetesEndpointController) identifyIPSubsetToAddOrUpdate(existingIPEndpointAddress map[string]corev1.EndpointAddress, desiredIPPod map[string]*corev1.Pod, fnCreateDesiredSubset func(pod *corev1.Pod, storageIP string) corev1.EndpointSubset) (ipSubsetToAdd map[string]corev1.EndpointSubset, ipSubsetToUpdate map[string]corev1.EndpointSubset) {
+	ipSubsetToAdd = make(map[string]corev1.EndpointSubset)
+	ipSubsetToUpdate = make(map[string]corev1.EndpointSubset)
+
+	for desiredIP, pod := range desiredIPPod {
+		desiredSubset := fnCreateDesiredSubset(pod, desiredIP)
+
+		// Check if the desired IP exists in the existing IP endpoint address.
+		if endpointAddress, exists := existingIPEndpointAddress[desiredIP]; exists {
+			// If it exists and the TargetRef is different, mark it for update.
```

Comment:
I don't think TargetRef is used with Kubernetes headless services. This is more relevant to Longhorn management for the Endpoints, as it helps identify which pods are currently backing the Service. It's a precautionary measure because IP assignments are managed by various CNI network plugins chosen by the user. This ensures the subset matches expectations, given that the headless Services are not managed by the kube-poxy.

> Is it expected that the address might be the same, but the TargetRef has changed? What can lead to this? (The share manager pod is recreated, but whatever handles the addressing for the storage network assigned the new pod the same IP address?)

I haven't encountered this scenario during my testing, but I have only tried a limited number of CNI plugin and IPAM combinations.